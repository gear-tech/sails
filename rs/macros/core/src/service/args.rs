use crate::sails_paths;
use proc_macro_error::abort;
use syn::{
    bracketed,
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    token, Path, Result as SynResult, Token,
};

#[derive(PartialEq, Debug)]
pub(super) struct ServiceArgs {
    base_types: Vec<Path>,
    events_type: Option<Path>,
    sails_path: Option<Path>,
}

impl ServiceArgs {
    pub fn base_types(&self) -> &[Path] {
        &self.base_types
    }

    pub fn events_type(&self) -> &Option<Path> {
        &self.events_type
    }

    pub fn sails_path(&self) -> syn::Path {
        sails_paths::sails_path_or_default(self.sails_path.clone())
    }
}

impl Parse for ServiceArgs {
    fn parse(input: ParseStream) -> SynResult<Self> {
        let items = input.parse_terminated(ServiceArg::parse, Token![,])?;
        let base_types = items
            .iter()
            .filter_map(|arg| match arg {
                ServiceArg::Extends(paths) => Some(paths.clone()),
                _ => None,
            })
            .flatten()
            .collect();
        let mut events_types = items.iter().filter_map(|arg| match arg {
            ServiceArg::Events(path) => Some(path.clone()),
            _ => None,
        });
        let events_type = events_types.next();
        if let Some(path) = events_types.next() {
            abort!(path, "only one `events` argument is allowed")
        }
        let sails_path = items.iter().find_map(|arg| match arg {
            ServiceArg::SailsPath(path) => Some(path.clone()),
            _ => None,
        });
        Ok(Self {
            base_types,
            events_type,
            sails_path,
        })
    }
}

#[derive(Debug)]
enum ServiceArg {
    Extends(Vec<Path>),
    Events(Path),
    SailsPath(Path),
}

impl Parse for ServiceArg {
    fn parse(input: ParseStream) -> SynResult<Self> {
        let path = input.parse::<Path>()?;
        let ident = path.get_ident().unwrap();
        input.parse::<Token![=]>()?;
        match ident.to_string().as_str() {
            "extends" => {
                if let Ok(path) = input.parse::<Path>() {
                    // Check path_expr.attrs is empty and qself is none
                    return Ok(Self::Extends(vec![path]));
                } else if input.peek(token::Bracket) {
                    let content;
                    let _bracket = bracketed!(content in input);
                    let punctuated: Punctuated<Path, Token![,]> =
                        Punctuated::parse_terminated(&content)?;
                    return Ok(Self::Extends(punctuated.into_iter().collect::<Vec<_>>()));
                }
                abort!(ident, "unexpected value for `extends` argument: {}", input)
            }
            "events" => {
                if let Ok(path) = input.parse::<Path>() {
                    return Ok(Self::Events(path));
                }
                abort!(ident, "unexpected value for `events` argument: {}", input)
            }
            "crate" => {
                if let Ok(path) = input.parse::<Path>() {
                    return Ok(Self::SailsPath(path));
                }
                abort!(ident, "unexpected value for `crate` argument: {}", input)
            }
            _ => abort!(ident, "unknown argument: {}", ident),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use proc_macro2::Span;
    use quote::quote;
    use syn::{
        punctuated::Punctuated, AngleBracketedGenericArguments, GenericArgument, Ident, Lifetime,
        PathArguments, PathSegment, Token,
    };

    #[test]
    fn gservice_parse_empty() {
        // arrange
        let input = quote!();

        let expected = ServiceArgs {
            base_types: vec![],
            events_type: None,
            sails_path: None,
        };

        // act
        let args = syn::parse2::<ServiceArgs>(input).unwrap();

        // arrange
        assert_eq!(expected, args);
    }

    #[test]
    fn gservice_parse_extends_path() {
        // arrange
        let input = quote!(extends = SomeService);

        let expected = ServiceArgs {
            base_types: vec![
                PathSegment::from(Ident::new("SomeService", Span::call_site())).into(),
            ],
            events_type: None,
            sails_path: None,
        };

        // act
        let args = syn::parse2::<ServiceArgs>(input).unwrap();

        // arrange
        assert_eq!(expected, args);
    }

    #[test]
    fn gservice_parse_extends_array_path() {
        // arrange
        let input = quote!(extends = [SomeService, AnotherService]);

        let expected = ServiceArgs {
            base_types: vec![
                PathSegment::from(Ident::new("SomeService", Span::call_site())).into(),
                PathSegment::from(Ident::new("AnotherService", Span::call_site())).into(),
            ],
            events_type: None,
            sails_path: None,
        };

        // act
        let args = syn::parse2::<ServiceArgs>(input).unwrap();

        // arrange
        assert_eq!(expected, args);
    }

    #[test]
    fn gservice_parse_extends_path_with_args() {
        // arrange
        let input = quote!(extends = SomeService<'a>);

        let lt = Lifetime::new("'a", Span::call_site());
        let mut args = Punctuated::new();
        args.push(GenericArgument::Lifetime(lt));
        let arguments = AngleBracketedGenericArguments {
            colon2_token: None,
            lt_token: Token![<](Span::call_site()),
            args,
            gt_token: Token![>](Span::call_site()),
        };

        let expected = ServiceArgs {
            base_types: vec![PathSegment {
                ident: Ident::new("SomeService", Span::call_site()),
                arguments: PathArguments::AngleBracketed(arguments),
            }
            .into()],
            events_type: None,
            sails_path: None,
        };

        // act
        let args = syn::parse2::<ServiceArgs>(input).unwrap();

        // arrange
        assert_eq!(expected, args);
    }

    #[test]
    fn gservice_parse_extends_array_path_with_args() {
        // arrange
        let input = quote!(extends = [BaseService, SomeService<'a>]);

        let lt = Lifetime::new("'a", Span::call_site());
        let mut args = Punctuated::new();
        args.push(GenericArgument::Lifetime(lt));
        let arguments = AngleBracketedGenericArguments {
            colon2_token: None,
            lt_token: Token![<](Span::call_site()),
            args,
            gt_token: Token![>](Span::call_site()),
        };

        let expected = ServiceArgs {
            base_types: vec![
                PathSegment::from(Ident::new("BaseService", Span::call_site())).into(),
                PathSegment {
                    ident: Ident::new("SomeService", Span::call_site()),
                    arguments: PathArguments::AngleBracketed(arguments),
                }
                .into(),
            ],
            events_type: None,
            sails_path: None,
        };

        // act
        let args = syn::parse2::<ServiceArgs>(input).unwrap();

        // arrange
        assert_eq!(expected, args);
    }
}
