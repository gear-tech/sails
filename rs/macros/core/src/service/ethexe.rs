use super::*;
use proc_macro2::TokenStream;
use quote::quote;

impl ServiceBuilder<'_> {
    pub(super) fn service_signature_impl(&self) -> TokenStream {
        let sails_path = self.sails_path;
        let service_type_path = self.type_path;
        let generics = &self.generics;
        let service_type_constraints = self.type_constraints.as_ref();
        let service_method_routes = self
            .service_handlers
            .iter()
            .map(|fn_builder| fn_builder.sol_handler_signature());

        let combined_methods = if self.base_types.is_empty() {
            quote! {
                &[#( #service_method_routes )*]
            }
        } else {
            let base_methods = self.base_types.iter().map(|path| {
                let path_wo_lifetimes = shared::remove_lifetimes(path);
                quote! {
                    <#path_wo_lifetimes as #sails_path::solidity::ServiceSignature>::METHODS
                }
            });
            quote! {
                #sails_path::const_concat_slices!(
                    <#sails_path::solidity::MethodRoute>,
                    &[#( #service_method_routes )*],
                    #( #base_methods ),*
                )
            }
        };

        quote! {
            impl #generics #sails_path::solidity::ServiceSignature for #service_type_path #service_type_constraints {
                const METHODS: &'static [#sails_path::solidity::MethodRoute] = #combined_methods;
            }
        }
    }

    pub(super) fn try_handle_solidity_impl(&self, base_ident: &Ident) -> TokenStream {
        let service_method_branches = self
            .service_handlers
            .iter()
            .map(|fn_builder| fn_builder.sol_try_handle_branch_impl());
        let base_types_try_handle = self.base_types.iter().enumerate().map(|(idx, _)| {
            let idx = Literal::usize_unsuffixed(idx);
            quote! {
                if let Some((output, value)) = self. #base_ident . #idx .try_handle_solidity(method, input).await {
                    return Some((output, value));
                }
            }
        });

        quote! {
            pub async fn try_handle_solidity(
                &mut self,
                method: &[u8],
                input: &[u8],
            ) -> Option<(Vec<u8>, u128)> {
                #( #service_method_branches )*
                #( #base_types_try_handle )*
                None
            }
        }
    }
}

impl FnBuilder<'_> {
    /// Generates code
    /// ```rust
    /// if method == &[24u8, 68u8, 111u8, 84u8, 104u8, 105u8, 115u8] {
    ///     // invocation
    /// }
    /// ```
    fn sol_try_handle_branch_impl(&self) -> TokenStream {
        let sails_path = self.sails_path;
        let handler_route_bytes = self.encoded_route.as_slice();
        let invocation = self.sol_invocation_func();

        quote! {
            if method == &[ #(#handler_route_bytes),* ] {
                #invocation
                return Some((#sails_path::alloy_sol_types::SolValue::abi_encode(&result), value));
            }
        }
    }

    /// Generates code for encode/decode parameters and fn invocation
    /// ```rust
    /// let (_, p1, p2): (u128, u32, String) = sails_rs::alloy_sol_types::SolValue::abi_decode_params(input, false).ok()?;
    /// let result: u32 = self.do_this(p1, p2).await;
    /// let value = 0u128;
    /// ```
    fn sol_invocation_func(&self) -> TokenStream {
        let sails_path = self.sails_path;
        let handler_ident = self.ident;
        let handler_params = self.params_idents();
        let handler_types = self.params_types();

        let (result_type, reply_with_value) = self.result_type_with_value();

        let await_token = self.is_async().then(|| quote!(.await));
        let unwrap_token = self.unwrap_result.then(|| quote!(.unwrap()));

        let handle_token = if reply_with_value {
            quote! {
                let command_reply: CommandReply< #result_type > = self.#handler_ident(#(#handler_params),*)#await_token #unwrap_token.into();
                let (result, value) = command_reply.to_tuple();
            }
        } else {
            quote! {
                let result = self.#handler_ident(#(#handler_params),*)#await_token #unwrap_token;
                let value = 0u128;
            }
        };

        quote! {
            let (_, #(#handler_params,)*) : (u128, #(#handler_types,)*) = #sails_path::alloy_sol_types::SolValue::abi_decode_params(input, false).ok()?;
            #handle_token
        }
    }
}
