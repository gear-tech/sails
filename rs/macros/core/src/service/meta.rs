use super::*;
use proc_macro2::TokenStream;
use quote::quote;

impl ServiceBuilder<'_> {
    pub(super) fn meta_trait_impl(&self, meta_module_ident: &Ident) -> TokenStream {
        let sails_path = self.sails_path;
        let generics = &self.generics;
        let service_type_path = self.type_path;
        let service_type_constraints = self.type_constraints();

        let base_services_meta = self.base_types.iter().map(|base_type| {
            quote! {
                #sails_path::meta::AnyServiceMeta::new::< #base_type >
            }
        });

        quote! {
            impl #generics #sails_path::meta::ServiceMeta for #service_type_path #service_type_constraints {
                type CommandsMeta = #meta_module_ident::CommandsMeta;
                type QueriesMeta = #meta_module_ident::QueriesMeta;
                type EventsMeta = #meta_module_ident::EventsMeta;
                const BASE_SERVICES: &'static [#sails_path::meta::AnyServiceMetaFn] = &[
                    #( #base_services_meta ),*
                ];
            }
        }
    }

    pub(super) fn meta_module(&self, meta_module_ident: &Ident) -> TokenStream {
        let sails_path = self.sails_path;
        let scale_codec_path = &sails_paths::scale_codec_path(sails_path);
        let scale_info_path = &sails_paths::scale_info_path(sails_path);

        let no_events_type = Path::from(Ident::new("NoEvents", Span::call_site()));
        let events_type = self.events_type.unwrap_or(&no_events_type);

        let invocation_params_structs = self
            .service_handlers
            .iter()
            .map(|fn_builder| fn_builder.params_struct(scale_codec_path, scale_info_path));
        let commands_meta_variants = self.service_handlers.iter().filter_map(|fn_builder| {
            (!fn_builder.is_query()).then_some(fn_builder.handler_meta_variant())
        });
        let queries_meta_variants = self.service_handlers.iter().filter_map(|fn_builder| {
            (fn_builder.is_query()).then_some(fn_builder.handler_meta_variant())
        });

        quote! {
            mod #meta_module_ident {
                use super::*;
                use #sails_path::{Decode, TypeInfo};

                #( #invocation_params_structs )*

                #[derive(TypeInfo)]
                #[scale_info(crate = #scale_info_path)]
                pub enum CommandsMeta {
                    #(#commands_meta_variants),*
                }

                #[derive(TypeInfo)]
                #[scale_info(crate = #scale_info_path)]
                pub enum QueriesMeta {
                    #(#queries_meta_variants),*
                }

                #[derive(TypeInfo)]
                #[scale_info(crate = #scale_info_path )]
                pub enum #no_events_type {}

                pub type EventsMeta = #events_type;
            }
        }
    }
}
