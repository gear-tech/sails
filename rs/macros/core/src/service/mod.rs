// This file is part of Gear.

// Copyright (C) 2021-2023 Gear Technologies Inc.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

//! Supporting functions and structures for the `gservice` macro.

use crate::{
    sails_paths,
    shared::{self, Func},
};
use args::ServiceArgs;
use convert_case::{Case, Casing};
use parity_scale_codec::Encode;
use proc_macro2::{Span, TokenStream};
use proc_macro_error::abort;
use quote::quote;
use std::collections::BTreeMap;
use syn::{
    parse_quote, punctuated::Punctuated, spanned::Spanned, token::Comma, Ident, ImplItemFn,
    ItemImpl, Lifetime, Path, Type, Visibility,
};

mod args;

static mut SERVICE_SPANS: BTreeMap<String, Span> = BTreeMap::new();

pub fn gservice(args: TokenStream, service_impl: TokenStream) -> TokenStream {
    let service_impl = parse_gservice_impl(service_impl);
    ensure_single_gservice_on_impl(&service_impl);
    ensure_single_gservice_by_name(&service_impl);
    generate_gservice(args, service_impl)
}

#[doc(hidden)]
pub fn __gservice_internal(args: TokenStream, service_impl: TokenStream) -> TokenStream {
    let service_impl = parse_gservice_impl(service_impl);
    generate_gservice(args, service_impl)
}

fn parse_gservice_impl(service_impl_tokens: TokenStream) -> ItemImpl {
    syn::parse2(service_impl_tokens).unwrap_or_else(|err| {
        abort!(
            err.span(),
            "`service` attribute can be applied to impls only: {}",
            err
        )
    })
}

fn ensure_single_gservice_on_impl(service_impl: &ItemImpl) {
    let attr_gservice = service_impl.attrs.iter().find(|attr| {
        attr.meta
            .path()
            .segments
            .last()
            .map(|s| s.ident == "service")
            .unwrap_or(false)
    });
    if attr_gservice.is_some() {
        abort!(
            service_impl,
            "multiple `service` attributes on the same impl are not allowed",
        )
    }
}

fn ensure_single_gservice_by_name(service_impl: &ItemImpl) {
    let (path, ..) = shared::impl_type(service_impl);
    let type_ident = path.path.segments.last().unwrap().ident.to_string();
    if unsafe { SERVICE_SPANS.get(&type_ident) }.is_some() {
        abort!(
            service_impl,
            "multiple `service` attributes on a type with the same name are not allowed"
        )
    }
    unsafe { SERVICE_SPANS.insert(type_ident, service_impl.span()) };
}

fn generate_gservice(args: TokenStream, service_impl: ItemImpl) -> TokenStream {
    let service_args = syn::parse2::<ServiceArgs>(args).unwrap_or_else(|err| {
        abort!(
            err.span(),
            "failed to parse `service` attribute arguments: {}",
            err
        )
    });
    let sails_path = service_args.sails_path();
    let scale_codec_path = sails_paths::scale_codec_path(&sails_path);
    let scale_info_path = sails_paths::scale_info_path(&sails_path);

    let (service_type_path, service_type_args) = shared::impl_type(&service_impl);
    let service_type_constraints = shared::impl_constraints(&service_impl);

    let service_handlers = discover_service_handlers(&service_impl);

    if service_handlers.is_empty() && service_args.base_types().is_empty() {
        abort!(
            service_impl,
            "`service` attribute requires impl to define at least one public method or extend another service"
        );
    }

    let events_type = service_args.events_type().as_ref();

    let mut service_impl = service_impl.clone();
    if let Some(events_type) = events_type {
        service_impl.items.push(parse_quote!(
            fn notify_on(&mut self, event: #events_type ) -> #sails_path::errors::Result<()>  {
                #[cfg(not(target_arch = "wasm32"))]
                {
                    let self_ptr = self as *const _ as usize;
                    let event_listeners = event_listeners().lock();
                    if let Some(event_listener_ptr) = event_listeners.get(&self_ptr) {
                        let event_listener =
                            unsafe { &mut *(*event_listener_ptr as *mut Box<dyn FnMut(& #events_type )>) };
                        core::mem::drop(event_listeners);
                        event_listener(&event);
                    }
                }
                #sails_path::gstd::events::__notify_on(event)
            }
        ));
    }

    let inner_ident = Ident::new("inner", Span::call_site());
    let inner_ptr_ident = Ident::new("inner_ptr", Span::call_site());
    let input_ident = Ident::new("input", Span::call_site());

    let mut exposure_funcs = Vec::with_capacity(service_handlers.len());
    let mut invocation_params_structs = Vec::with_capacity(service_handlers.len());
    let mut invocation_funcs = Vec::with_capacity(service_handlers.len());
    let mut invocation_dispatches = Vec::with_capacity(service_handlers.len());
    let mut commands_meta_variants = Vec::with_capacity(service_handlers.len());
    let mut queries_meta_variants = Vec::with_capacity(service_handlers.len());

    for (handler_route, (handler_fn, ..)) in &service_handlers {
        let handler_fn = &handler_fn.sig;
        let handler_func = Func::from(handler_fn);
        let handler_generator = HandlerGenerator::from(handler_func.clone());
        let invocation_func_ident = handler_generator.invocation_func_ident();

        exposure_funcs.push({
            let handler_ident = handler_func.ident();
            let handler_params = handler_func.params().iter().map(|item| item.0);
            let handler_await_token = handler_func.is_async().then(|| quote!(.await));
            quote!(
                pub #handler_fn {
                    let exposure_scope = #sails_path::gstd::services::ExposureCallScope::new(self);
                    self. #inner_ident . #handler_ident (#(#handler_params),*) #handler_await_token
                }
            )
        });
        invocation_params_structs.push(handler_generator.params_struct());
        invocation_funcs.push(handler_generator.invocation_func());
        invocation_dispatches.push({
            let handler_route_bytes = handler_route.encode();
            let handler_route_len = handler_route_bytes.len();
            quote!(
                if #input_ident.starts_with(& [ #(#handler_route_bytes),* ]) {
                    let output = self.#invocation_func_ident(&#input_ident[#handler_route_len..]).await;
                    static INVOCATION_ROUTE: [u8; #handler_route_len] = [ #(#handler_route_bytes),* ];
                    return Some([INVOCATION_ROUTE.as_ref(), &output].concat());
                }
            )
        });

        let handler_meta_variant = {
            let params_struct_ident = handler_generator.params_struct_ident();
            let result_type = handler_generator.result_type();
            let handler_route_ident = Ident::new(handler_route, Span::call_site());

            quote!(#handler_route_ident(#params_struct_ident, #result_type))
        };
        if handler_generator.is_query() {
            queries_meta_variants.push(handler_meta_variant);
        } else {
            commands_meta_variants.push(handler_meta_variant);
        }
    }

    let message_id_ident = Ident::new("message_id", Span::call_site());
    let route_ident = Ident::new("route", Span::call_site());

    let code_for_base_types = service_args.base_types().iter()
        .enumerate()
        .map(|(idx, base_type)| {
            let base_ident = Ident::new(&format!("base_{}", idx), Span::call_site());
            let as_base_ident = Ident::new(&format!("as_base_{}", idx), Span::call_site());

            let base_exposure_accessor = quote!(
                pub fn #as_base_ident (&self) -> &< #base_type as #sails_path::gstd::services::Service>::Exposure {
                    &self. #base_ident
                }
            );

            let base_exposure_member = quote!(
                #base_ident : < #base_type as #sails_path::gstd::services::Service>::Exposure,
            );

            let base_exposure_instantiation = quote!(
                #base_ident : < #base_type as Clone>::clone(AsRef::< #base_type >::as_ref( #inner_ident ))
                    .expose( #message_id_ident , #route_ident ),
            );

            let base_exposure_invocation = quote!(
                if let Some(output) = self. #base_ident .try_handle(#input_ident).await {
                    return Some(output);
                }
            );

            let base_service_meta = quote!(#sails_path::meta::AnyServiceMeta::new::< #base_type >());

            (base_exposure_accessor, base_exposure_member, base_exposure_instantiation, base_exposure_invocation, base_service_meta)
        });

    let base_exposure_accessors = code_for_base_types
        .clone()
        .map(|(base_exposure_accessor, ..)| base_exposure_accessor);

    let base_exposures_members = code_for_base_types
        .clone()
        .map(|(_, base_exposure_member, ..)| base_exposure_member);

    let base_exposures_instantiations = code_for_base_types
        .clone()
        .map(|(_, _, base_exposure_instantiation, ..)| base_exposure_instantiation);

    let base_exposures_invocations = code_for_base_types
        .clone()
        .map(|(_, _, _, base_exposure_invocation, ..)| base_exposure_invocation);

    let base_services_meta =
        code_for_base_types.map(|(_, _, _, _, base_service_meta)| base_service_meta);

    let events_listeners_code = events_type.map(|_| generate_event_listeners(&sails_path));

    let lifetimes = shared::extract_lifetime_names(&service_type_args);
    let exposure_set_event_listener_code = events_type.map(|t| {
        // get non conflicting lifetime name

        let mut lt = "__elg".to_owned();
        while lifetimes.contains(&lt) {
            lt = format!("_{}", lt);
        }
        let lifetime_name = format!("'{0}", lt);
        generate_exposure_set_event_listener(t, Lifetime::new(&lifetime_name, Span::call_site()))
    });

    let exposure_drop_code = events_type.map(|_| generate_exposure_drop());

    let no_events_type = Path::from(Ident::new("NoEvents", Span::call_site()));
    let events_type = events_type.unwrap_or(&no_events_type);

    let unexpected_route_panic =
        shared::generate_unexpected_input_panic(&input_ident, "Unknown request", &sails_path);

    let mut exposure_lifetimes: Punctuated<Lifetime, Comma> = Punctuated::new();
    if !service_args.base_types().is_empty() {
        for lt in lifetimes.iter().map(|lt| {
            let lt = format!("'{lt}");
            Lifetime::new(&lt, Span::call_site())
        }) {
            exposure_lifetimes.push(lt);
        }
    };

    let exposure_generic_args = if exposure_lifetimes.is_empty() {
        quote! { T }
    } else {
        quote! { #exposure_lifetimes, T }
    };
    let exposure_args = if exposure_lifetimes.is_empty() {
        quote! { #service_type_path }
    } else {
        quote! { #exposure_lifetimes, #service_type_path }
    };

    quote!(
        #service_impl

        pub struct Exposure<#exposure_generic_args> {
            #message_id_ident : #sails_path::MessageId,
            #route_ident : &'static [u8],
            #[cfg(not(target_arch = "wasm32"))]
            #inner_ident : Box<T>, // Ensure service is not movable
            #[cfg(not(target_arch = "wasm32"))]
            #inner_ptr_ident : *const T, // Prevent exposure being Send + Sync
            #[cfg(target_arch = "wasm32")]
            #inner_ident : T,
            #( #base_exposures_members )*
        }

        #exposure_drop_code

        impl #service_type_args Exposure< #exposure_args > #service_type_constraints {
            #( #exposure_funcs )*

            #( #base_exposure_accessors )*

            pub async fn handle(&mut self, #input_ident: &[u8]) -> Vec<u8> {
                self.try_handle( #input_ident ).await.unwrap_or_else(|| {
                    #unexpected_route_panic
                })
            }

            pub async fn try_handle(&mut self, #input_ident : &[u8]) -> Option<Vec<u8>> {
                #( #invocation_dispatches )*
                #( #base_exposures_invocations )*
                None
            }

            #( #invocation_funcs )*

            #exposure_set_event_listener_code
        }

        impl #service_type_args #sails_path::gstd::services::Exposure for Exposure< #exposure_args > #service_type_constraints {
            fn message_id(&self) -> #sails_path::MessageId {
                self. #message_id_ident
            }

            fn route(&self) -> &'static [u8] {
                self. #route_ident
            }
        }

        impl #service_type_args #sails_path::gstd::services::Service for #service_type_path #service_type_constraints {
            type Exposure = Exposure< #exposure_args >;

            fn expose(self, #message_id_ident : #sails_path::MessageId, #route_ident : &'static [u8]) -> Self::Exposure {
                #[cfg(not(target_arch = "wasm32"))]
                let inner_box = Box::new(self);
                #[cfg(not(target_arch = "wasm32"))]
                let #inner_ident = inner_box.as_ref();
                #[cfg(target_arch = "wasm32")]
                let #inner_ident = &self;
                Self::Exposure {
                    #message_id_ident ,
                    #route_ident ,
                    #( #base_exposures_instantiations )*
                    #[cfg(not(target_arch = "wasm32"))]
                    #inner_ptr_ident : inner_box.as_ref() as *const Self,
                    #[cfg(not(target_arch = "wasm32"))]
                    #inner_ident : inner_box ,
                    #[cfg(target_arch = "wasm32")]
                    #inner_ident : self,
                }
            }
        }

        impl #service_type_args #sails_path::meta::ServiceMeta for #service_type_path #service_type_constraints {
            fn commands() -> #scale_info_path ::MetaType {
                #scale_info_path ::MetaType::new::<meta_in_service::CommandsMeta>()
            }

            fn queries() -> #scale_info_path ::MetaType {
                #scale_info_path ::MetaType::new::<meta_in_service::QueriesMeta>()
            }

            fn events() -> #scale_info_path ::MetaType {
                #scale_info_path ::MetaType::new::<meta_in_service::EventsMeta>()
            }

            fn base_services() -> impl Iterator<Item = #sails_path::meta::AnyServiceMeta> {
                [
                    #( #base_services_meta ),*
                ].into_iter()
            }
        }

        #events_listeners_code

        use #sails_path ::Decode as __ServiceDecode;
        use #sails_path ::Encode as __ServiceEncode;
        use #sails_path ::TypeInfo as __ServiceTypeInfo;

        #(
            #[derive(__ServiceDecode, __ServiceTypeInfo)]
            #[codec(crate = #scale_codec_path )]
            #[scale_info(crate = #scale_info_path )]
            #invocation_params_structs
        )*

        mod meta_in_service {
            use super::*;

            #[derive(__ServiceTypeInfo)]
            #[scale_info(crate = #scale_info_path )]
            pub enum CommandsMeta {
                #(#commands_meta_variants),*
            }

            #[derive(__ServiceTypeInfo)]
            #[scale_info(crate = #scale_info_path )]
            pub enum QueriesMeta {
                #(#queries_meta_variants),*
            }

            #[derive(__ServiceTypeInfo)]
            #[scale_info(crate = #scale_info_path )]
            pub enum #no_events_type {}

            pub type EventsMeta = #events_type;
        }
    )
}

// Generates function for accessing event listeners map in non-wasm code.
fn generate_event_listeners(sails_path: &Path) -> TokenStream {
    quote!(
        type __EventlistenersMap = #sails_path::collections::BTreeMap<usize, usize>;
        type __Mutex<T> = #sails_path::spin::Mutex<T>;

        #[cfg(not(target_arch = "wasm32"))]
        fn event_listeners() -> &'static __Mutex<__EventlistenersMap> {
            static EVENT_LISTENERS: __Mutex<__EventlistenersMap> =
                __Mutex::new(__EventlistenersMap::new());
            &EVENT_LISTENERS
        }

        #[cfg(not(target_arch = "wasm32"))]
        pub struct EventListenerGuard<'a> {
            service_ptr: usize,
            listener_ptr: usize,
            _phantom: core::marker::PhantomData<&'a ()>,
        }

        #[cfg(not(target_arch = "wasm32"))]
        impl<'a> Drop for EventListenerGuard<'a> {
            fn drop(&mut self) {
                let mut event_listeners = event_listeners().lock();
                let listener_ptr = event_listeners.remove(&self.service_ptr);
                if listener_ptr != Some(self.listener_ptr) {
                    panic!("event listener is being removed out of order");
                }
            }
        }
    )
}

fn generate_exposure_drop() -> TokenStream {
    quote!(
        #[cfg(not(target_arch = "wasm32"))]
        impl<T> Drop for Exposure<T> {
            fn drop(&mut self) {
                let service_ptr = self.inner_ptr as usize;
                let mut event_listeners = event_listeners().lock();
                if event_listeners.remove(&service_ptr).is_some() {
                    panic!("there should be no any event listeners left by this time");
                }
            }
        }
    )
}

fn generate_exposure_set_event_listener(events_type: &Path, lifetime: Lifetime) -> TokenStream {
    quote!(
        #[cfg(not(target_arch = "wasm32"))]
        // Immutable so one can set it via AsRef when used with extending
        pub fn set_event_listener<#lifetime>(
            &self,
            listener: impl FnMut(& #events_type ) + #lifetime,
        ) -> EventListenerGuard<#lifetime> {
            if core::mem::size_of_val(self.inner.as_ref()) == 0 {
                panic!("setting event listener on a zero-sized service is not supported for now");
            }
            let service_ptr = self.inner_ptr as usize;
            let listener: Box<dyn FnMut(& #events_type )> = Box::new(listener);
            let listener = Box::new(listener);
            let listener_ptr = Box::into_raw(listener) as usize;
            let mut event_listeners = event_listeners().lock();
            if event_listeners.contains_key(&service_ptr) {
                panic!("event listener is already set");
            }
            event_listeners.insert(service_ptr, listener_ptr);
            EventListenerGuard {
                service_ptr,
                listener_ptr,
                _phantom: core::marker::PhantomData,
            }
        }
    )
}

fn discover_service_handlers(service_impl: &ItemImpl) -> BTreeMap<String, (&ImplItemFn, usize)> {
    shared::discover_invocation_targets(service_impl, |fn_item| {
        matches!(fn_item.vis, Visibility::Public(_)) && fn_item.sig.receiver().is_some()
    })
}

struct HandlerGenerator<'a> {
    handler: Func<'a>,
}

impl<'a> HandlerGenerator<'a> {
    fn from(handler: Func<'a>) -> Self {
        Self { handler }
    }

    fn params_struct_ident(&self) -> Ident {
        Ident::new(
            &format!(
                "__{}Params",
                self.handler.ident().to_string().to_case(Case::Pascal)
            ),
            Span::call_site(),
        )
    }

    fn result_type(&self) -> Type {
        shared::replace_any_lifetime_with_static(self.handler.result().clone())
    }

    fn handler_func_ident(&self) -> Ident {
        self.handler.ident().clone()
    }

    fn invocation_func_ident(&self) -> Ident {
        Ident::new(
            &format!("__{}", self.handler_func_ident()),
            Span::call_site(),
        )
    }

    fn is_query(&self) -> bool {
        self.handler
            .receiver()
            .map_or(true, |r| r.mutability.is_none())
    }

    fn params_struct(&self) -> TokenStream {
        let params_struct_ident = self.params_struct_ident();
        let params_struct_members = self.handler.params().iter().map(|item| {
            let arg_ident = item.0;
            let arg_type = item.1;
            quote!(#arg_ident: #arg_type)
        });

        quote!(
            pub struct #params_struct_ident {
                #(#params_struct_members),*
            }
        )
    }

    fn invocation_func(&self) -> TokenStream {
        let invocation_func_ident = self.invocation_func_ident();
        let receiver = self.handler.receiver();
        let params_struct_ident = self.params_struct_ident();
        let handler_func_ident = self.handler_func_ident();
        let handler_func_params = self.handler.params().iter().map(|item| {
            let param_ident = item.0;
            quote!(request.#param_ident)
        });

        let await_token = self.handler.is_async().then(|| quote!(.await));

        quote!(
            async fn #invocation_func_ident(#receiver, mut input: &[u8]) -> Vec<u8>
            {
                let request = #params_struct_ident::decode(&mut input).expect("Failed to decode request");
                let result = self.#handler_func_ident(#(#handler_func_params),*)#await_token;
                return result.encode();
            }
        )
    }
}
