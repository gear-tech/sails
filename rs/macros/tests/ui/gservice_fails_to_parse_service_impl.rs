use sails_macros::service;

#[service]
struct MyService;

#[tokio::main]
async fn main() {}
