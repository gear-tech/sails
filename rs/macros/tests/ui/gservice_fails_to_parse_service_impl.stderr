error: `service` attribute can be applied to impls only: expected `impl`
 --> tests/ui/gservice_fails_to_parse_service_impl.rs:4:1
  |
4 | struct MyService;
  | ^^^^^^
